examples/experimental/dagster-airlift/dagster_airlift/core/state_backed_defs_loader.py

```diff
@@ -0,0 +1,127 @@
+
+    The goal here is so that an unreliable backing source is fetched only
+    at code server load time, which defines the code location. When, for example,
+    a run worker is launched, the backing source is not queryed again, and
+    only `defs_from_state` is called.
+
+    Current meant for internal usage only hence TState must be a marked as whitelist_for_serdes.
+
+    Args:
+        defs_key (str): The unique key for the definitions. Must be unique per code location.
+
+    Examples:
+
+    .. code-block:: python
+
+        @whitelist_for_serdes
+        @record
+        class ExampleDefState:
+            a_string: str
+
+        class ExampleStateBackedDefinitionsLoader(StateBackedDefinitionsLoader[ExampleDefState]):
+            def fetch_backing_state(self) -> ExampleDefState:
+                # Fetch from potentially unreliable API (e.g. Rest API)
+                return ExampleDefState(a_string="foo")
+
+            def defs_from_state(self, state: ExampleDefState) -> Definitions:
+                # Construct or reconstruct the Definitions from the previously
+                # fetched state.
+                return Definitions([AssetSpec(key=state.a_string)])
+    """
+
+    @property
+    @abstractmethod
+    def defs_key(self) -> str:
+        """The unique key for the definitions. Must be unique per code location."""
+        ...
+
+    @abstractmethod
+    def fetch_backing_state(self) -> TState:
+        """Subclasses must implement this method. This is where the integration runs
+        code that fetches the backing state from the source of truth for the definitions.
+        This is only called when the code location is initializing, for example on
+        code server load, or when loading via dagster dev.
+        """
+        ...
+
+    @abstractmethod
+    def defs_from_state(self, state: TState) -> Definitions:
+        """Subclasses must implement this method. It is invoked whenever the code location
+        is loading, whether it be initializaton or reconstruction. In the case of
+        intialization, it takes the result of fetch_backing state that just happened.
+        When reconstructing, it takes the state that was previously fetched and attached
+        as metadata.
+
+        This method also takes responsibility for attaching the state to the definitions
+        on its metadata, with the key passed in as defs_key.
+        """
+        ...
+
+    def build_defs(self) -> Definitions:
+        context = DefinitionsLoadContext.get()
+        state = (
+            context.reconstruction_metadata[self.defs_key]
+            if context.load_type == DefinitionsLoadType.RECONSTRUCTION
+            and self.defs_key in context.reconstruction_metadata
+            else self.fetch_backing_state()
+        )
+        return self.defs_from_state(state).with_reconstruction_metadata({self.defs_key: state})
+
+
+@contextmanager
+def scoped_metadata_for_reconstruction(**state_objects: Any) -> Generator:
```

Comment:
We use just "reconstruction metadata" elsewhere-- wdyt of naming this `scoped_reconstruction_metadata`, `temp_reconstruction_metadata`, `set_reconstruction_metadata`?